functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,140 @@
+@contextmanager
+def suspend_mode(mode):
+    assert(mode is not None), "Cannot suspend None mode"
+    assert(isinstance(mode, TorchDispatchMode)), f"Unexpected mode type {mode.__class__}"
+    torch._C._set_torch_dispatch_mode(None)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(mode)
+
+@contextmanager
+def enable_mode(mode):
+    curr_mode = torch._C._get_torch_dispatch_mode()
+    torch._C._set_torch_dispatch_mode(mode)
+    try:
+        yield
+    finally:
+        torch._C._set_torch_dispatch_mode(curr_mode)
+
+
+def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
+    def _unwrap_proxy(e):
+        return get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy)
+
+    assert isinstance(operands, list), "Cond operands must be a list of tensors"
+    assert all(isinstance(o, torch.Tensor) for o in operands), "Cond operands must be a list of tensors"
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+
+    true_outs = []
+    false_outs = []
+    for node in true_graph.graph.nodes:
+        if node.op == 'output':
+            true_outs.extend(node.args)
+
+    for node in false_graph.graph.nodes:
+        if node.op == 'output':
+            false_outs.extend(node.args)
+
+    flat_true_outs, _ = pytree.tree_flatten(true_outs)
+    flat_false_outs, _ = pytree.tree_flatten(false_outs)
+    assert(len(flat_true_outs) == len(flat_false_outs))
+
+    for i in range(0, len(flat_true_outs)):
+        true_out = flat_true_outs[i]
+        false_out = flat_false_outs[i]
+        assert(true_out.meta == false_out.meta)
+
+    # There are probably better ways - I know that create_arg has some self incrementing name
+    # magic to it, but since we explicitly have to get the name for register_module,
+    # I was not sure how to do that. This kinda simulates it.
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    true_name = next_name
+    false_name = f"false_graph_{i}"
+    assert(not hasattr(proxy_mode.tracer.root, false_name))
+
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    args = (pred, true_graph, false_graph, [operands])
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    out_proxy = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, {},
+        name="conditional")
+
+    # We could get an out tensor by running the real ops
+    # But to avoid running real code in tracing, we just use a dummy tensor.
+    # if pred:
+    #     out = true_fn(*operands)
+    # else:
+    #     out = false_fn(*operands)
+    out = torch.zeros([])
```

Comment:
I'm not sure this is going to fly, if we're doing real tracing (`make_fx(f, mode='real')`), we end up executing real operations on the Tensor.

If we're tracing a function that calls cond and then passes the output of cond to another pytorch operation (like torch.sum), then `torch.sum` is going to receive this dummy Tensor and then will actually execute `out.sum()`. This will be a problem if the other operation is something that does rely on the output to be a real Tensor.

Here's an hacky idea that might work:
-  create `new_true_fn`. `new_true_fn` executes `true_fn` and then stores the output of `true_fn` somewhere.
```
saved_true_out = []
def new_true_fn(*operands):
  out = true_fn(operands)
  saved_out.append(out)
  return out
```
- Do `true_graph = get_isolated_graphmodule(new_true_fn, operands, {})` (and ditto for new_false_fn).
- Now, instead of `out = torch.zeros([])`, we can return either `out = saved_true_out[0]` or `out = saved_false_out[0]`.
- Furthermore, we can compare `saved_true_out` and `saved_false_out` for if "all the properties match" (though maybe that is what your checks above are doing, I'm not familiar with the code).

Reply:
We might as well just run the real thing without all the extra pizzaz around `new_true_fn` though. Just uncomment my commented code. 

Will change.

> Furthermore, we can compare saved_true_out and saved_false_out for if "all the properties match" (though maybe that is what your checks above are doing, I'm not familiar with the code).

We already do that above, yes.